cli/resolver.rs

```diff
@@ -26,6 +27,25 @@ use crate::npm::NpmResolution;
 use crate::npm::PackageJsonDepsInstaller;
 use crate::util::sync::AtomicFlag;
 
+/// Check that a resolved specifier isn't an `ext:` URL. Normally scheme checks
+/// are done on load rather than resolve. This is needed because `ext:` modules
+/// are preloaded in deno_core and won't be requested from the embedder, so we
+/// need to catch them on resolution instead.
+///
+/// TODO(nayeemrmn): Maybe use a separate module map for `ext:` modules to avoid
+/// this problem. As of writing this is blocked by the way `node:` specifiers
+/// are implemented.
+pub fn validate_scheme_for_resolution(
```

Comment:
Are we 100% sure there are no scenarios where rejecting on load is too late?

Also, any ideas how you'd handle `node:` specifiers? Maybe we should add an "alias" system for the `ext:` specifiers?

Reply:
Fair enough, any thoughts on `node:` specifiers?

Reply:
I figure we can store a list of aliases on the extension. My first way was to store that information on each extension module source / module map entry, but this way doesn't have to preserve it through snapshot layers since extension metadata is passed at every step.

Reply:
I thought of another approach. All of the ideas we've discussed so far require `deno_core` to guess at what the user wants to do with the extension module map based on whether or not a snapshot will be created, and are cryptically specific to CLI's usage. Instead I think we should have explicit procedural bindings for those behaviours. They will be very niche but it's unavoidable because that's the granularity we need:
- Keep all extension modules in the module map by default.
- Add the following APIs to `JsRuntime`:
  ```rust
  impl JsRuntime {
    /// Clear the internal module map.
    fn clear_module_map(&self) {}
    /// Get a module handle by specifier from the internal module map.
    fn get_module_handle(&self, specifier: &Url) -> Option<v8::Global<v8::Module>> {}
    /// Inject a module handle by specifier into the internal module map. The module must be
    /// kEvaluated. It will be stored with no dependency references.
    fn inject_module_handle(&self, specifier: Url, v8::Global<v8::Module>) {}
  }
  ```
- Create a runtime worker roughly like this:
  ```rust
  fn get_ext_node_module_handles(js_runtime: &JsRuntime) -> Vec<(Url, v8::Global<v8::Module>)> {
    vec![
      ("node:assert", js_runtime.get_module_handle("ext:deno_node/assert.ts").unwrap()),
      ("node:buffer", js_runtime.get_module_handle("ext:deno_node/buffer.ts").unwrap()),
      // ...
    ]
  }
  let js_runtime = JsRuntime::new(options);
  let ext_node_module_handles = get_ext_node_module_handles(&js_runtime);
  js_runtime.clear_Module_map();
  for (specifier, handle) in ext_node_module_handles {
    js_runtime.inject_module_handle(specifier, handle);
  }
  ```